urlbc3e98d97e_jmhorjus_EFCP_Implementation_tgz-pJ8-UDPConnector_UdpConnectorTestJ8/src/UDPConnector/EfcpConnector.java

```diff
@@ -0,0 +1,515 @@
+ * All rights reserved. Permission to use, copy, modify, and distribute this software and its documentation
+ * for any purpose and without fee is hereby granted, provided that the above copyright notice appear in all 
+ * copies and that both the copyright notice and this permission notice appear in supporting documentation. 
+ * The RINA laboratory of the Computer Science Department at Boston University makes no 
+ * representations about the suitability of this software for any purpose.
+ */
+package UDPConnector;
+
+import java.net.InetAddress;
+import java.util.ArrayList;
+import java.util.Date;
+import java.util.HashMap;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.Queue;
+import java.util.concurrent.ScheduledThreadPoolExecutor;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+
+/**
+ *
+ * @author Jan Horjus
+ */
+public class EfcpConnector implements ConnectorInterface
+{
+    /// The ConnectorInterface functions:
+    @Override
+    public void SetPeerAddress(
+            InetAddress peerAddress, 
+            int port)
+    {
+        m_innerConnection.SetPeerAddress(peerAddress, port);
+    }
+    @Override
+    public List<byte[]> Receive(int maxBlockingTimeInMs)
+    {
+        // This is for delivering data up to...delimiting, I think. It needs 
+        // to deliver a list of the buffers from the DtpPacket objects.         
+        List<byte[]> retVal = new ArrayList<>();
+        
+        synchronized(m_receiverPacketsReady)
+        {   
+            if (m_receiverPacketsReady.isEmpty())
+            {
+                System.out.print("Efcp.Receive: Nothing available; waiting.\n");
+                try{
+                    m_receiverPacketsReady.wait(maxBlockingTimeInMs);
+                } catch (Exception ex) { System.out.print("Efcp.Receive: ERROR:"+ex.getMessage()+".\n"); }
+            }
+            retVal.addAll(m_receiverPacketsReady);
+            m_receiverPacketsReady.clear();
+        }
+        
+        return retVal;
+    }
+    @Override
+    public boolean Send(String sendString) throws Exception
+    {
+        return Send(sendString.getBytes());
+    }
+    
+    protected boolean IsRateBasedFlowControlWindowOpen(){
+        Date now = new Date();
+        if (!m_policyInfo.RateFlowControlEnabled){
+            return true;
+        }
+        if(now.getTime() > m_senderRateCurrentPeriodStartTime.getTime() + m_policyInfo.RateDefaultPeriodInMs){
+            return true;
+        }
+        if(m_senderSendsSoFarThisPeriod < m_policyInfo.RateDefaultPaketsPerPeriod) {
+            return true;
+        }
+        System.out.print("Efcp: Rate based flow control engaged to close window.\n");
+        return false;
+    }
+    
+    protected void RateBasedFlowSend(){
+        if (!m_policyInfo.RateFlowControlEnabled){
+            return;
+        }
+        synchronized(m_senderRateCurrentPeriodStartTime)
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Object m_senderRateCurrentPeriodStartTime synchronized on line 83 should not be re-assigned in the same synchronized block on line 89. After this re-assignment, you lose the reference to the synchronized object.